File: LV 2/쿼드압축 후 개수 세기.py
import sys

count = [0, 0]

sys.setrecursionlimit(10**6)

def check(arr, x1, y1, x2, y2):
    t = arr[x1][y1]
    for i in range(x1, x2+1):
        for j in range(y1, y2+1):
            if arr[i][j] != t:
                return False
    
    return True

def f(arr, x1, y1, x2, y2):
    print(x1, y1, x2, y2)
    if check(arr, x1, y1, x2, y2):
        t1 = arr[x1][y1]
        count[t1] += 1
        return
    
    if x1 == x2 and y1 == y2:
        count[arr[x1][y1]] += 1
        return
    
    midX = (x1 + x2) // 2
    midY = (y1 + y2) // 2
    
    f(arr, x1, y1, midX, midY) # 왼위
    f(arr, midX+1, y1, x2, midY) # 왼아
    f(arr, midX+1, midY+1, x2, y2) # 오아
    f(arr, x1, midY+1, midX, y2) # 오위
    
    

def solution(arr):
    answer = []
    t = len(arr)
    f(arr, 0, 0, t-1, t-1)
    return count

File: LV 2/쿼드압축 후 개수 세기.java
class Solution {

    private int zeroCount = 0;
    private int oneCount = 0;

    public boolean check(int[][] arr, int sx, int ex, int sy, int ey) {
        int element = arr[sx][sy];
        for (int i = sx; i <= ex; i++) {
            for (int j = sy; j <= ey; j++) {
                if (arr[i][j] != element) {
                    return false;
                }
            }
        }
        return true;
    }

    public void f(int[][] arr, int sx, int ex, int sy, int ey) {
        if (check(arr, sx, ex, sy, ey)) {
            int data = arr[sx][sy];
            if (data == 1) {
                oneCount += 1;
            }

            if (data == 0) {
                zeroCount += 1;
            }

            return;
        }

        int midX = (sx + ex) / 2;
        int midY = (sy + ey) / 2;
        f(arr, sx, midX, sy, midY);
        f(arr, midX + 1, ex, sy, midY);
        f(arr, sx, midX, midY + 1, ey);
        f(arr, midX + 1, ex, midY + 1, ey);

    }

    public int[] solution(int[][] arr) {
        f(arr, 0, arr.length - 1, 0, arr[0].length - 1);
        int[] answer = {zeroCount, oneCount};
        return answer;
    }

}
